075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,368 @@
+
+The idea is about adding a new `autoRebalancing()` method in the `KafkaReconcile.reconcile()` flow, taking care of starting a rebalancing and checking the status of an ongoing one across reconciliations.
+The `KafkaReconcile.reconcile()` flow, taking into account methods involved in the scaling and rebalancing, would be something like:
+
+* ... 
+* `scaleDown()`
+* ...
+* `podSet()`
+* ...
+* `autoRebalancing()`
+* ...
+
+> In case of scaling down, the rebalancing has to happen before removing brokers and not after that as it looks like from the above flow. Despite that, some internal state will help to have the right actions order across reconciliations as explained in the next sections.
+
+#### Auto-rebalancing on scaling up
+
+The following flow is triggered when the user increases the `replicas` field within a `Kafka` or `KafkaNodePool` custom resource in order to scale up the cluster by adding more brokers.
+
+1. On the **1st** reconciliation (scaling up is requested): 
+     * the `podSet()` method triggers the scaling up, as described in the previous section, but the reconciliation is not blocked on waiting for the newly added pods to be ready.
+     * the `autoRebalancing()` method detects that a scaling up was requested because of `KafkaCluster.addedNodes()` returning a non empty list with the brokers' IDs to be added to the cluster. Because the `Kafka.status.cruiseControl.autoRebalance.state` doesn't exist, it's a new operation so the state is updated as `ScalingUp` and the `add-brokers` with the brokers' IDs. Of course, it doesn't trigger any rebalance because the scaling up is going on asynchronously.
+2. From the **2nd** to the **(N-1)th** reconciliation (scaling up is going on, checks when it's done):
+     * the `podSet()` does nothing because no new brokers were added this time.
+     * the `autoRebalancing()` evaluates the `Kafka.status.cruiseControl.autoRebalance.state: ScalingUp`, so a scaling up is running, and it has to check if it's finished or not. This can be achieved looking at the involved `StrimziPodSet`(s) and comparing `readyPods` and `pods` fields in the corresponding status. If they differ, the scaling up is still running. It should not take long time but it could go across multiple reconciliation, so this check repeats.
+3. On the **Nth** reconciliation (scaling up is done, rebalancing can start):
+     * the `podSet()` does nothing because no new brokers are added this time.
+     * the `autoRebalancing()` evaluates the `Kafka.status.cruiseControl.autoRebalance.state: ScalingUp`, so a scaling up is running, but finally looking at the involved `StrimziPodSet`(s), it gets `readyPods` equals to `pods` so the scaling up is done. A new `KafkaRebalance` custom resource is created and the `Kafka.status.cruiseControl.autoRebalance.state` is updated with the corresponding state (should be `ProposalPending`). The `KafkaRebalanceAssemblyOperator` takes care of handling the rebalancing.
```

Comment:
How will you name the rebalance resource? How do you ensure uniqueness and deterministic name?

Reply:
It's explained around line 115, it will be `<my-cluster-name>-auto-rebalancing-<mode>`.